pkg/controller/common/api/v2/common.go

```diff
@@ -0,0 +1,147 @@
+// it holds entries in form [GroupVersionKind] -> [namespace/name] -> object
+// where namespace is empty in case of resources which are cluster scoped,
+// i.e. :
+// "v1/Pod -> 'prometheus/Prometheus'" (for namespaced child resource if the parent is cluster scope)
+// "v1/Pod -> 'prometheus/Prometheus'" (for namespaced child resource if parent is namespaced)
+// "v1/Namespace -> 'some'" (for cluster-scope child resource if parent is cluster scope)
+type UniformObjectMap map[GroupVersionKind]map[string]*unstructured.Unstructured
+
+// MakeUniformObjectMap builds the map of objects resources that is suitable for use
+// in the `children` field of a CompositeController SyncRequest or
+// `attachments` field of  the  DecoratorControllers SyncRequest or `customize` field of
+// customize hook
+//
+// This function returns a UniformObjectMap which is a map of maps. The outer most map
+// is keyed  using the object's type and the inner map is keyed using the
+// object's namespace/name. If the object is clustered scope it it just its name
+func MakeUniformObjectMap(parent v1.Object, list []*unstructured.Unstructured) UniformObjectMap {
+	objectMap := make(UniformObjectMap)
+
+	objectMap.InsertAll(parent, list)
+
+	return objectMap
+}
+
+// InitGroup initializes a map for given schema.GroupVersionKind if not yet initialized
+func (m UniformObjectMap) InitGroup(gvk schema.GroupVersionKind) {
+	internalGvk := GroupVersionKind{gvk}
+	if m[internalGvk] == nil {
+		m[internalGvk] = make(map[string]*unstructured.Unstructured)
+	}
+}
+
+// Insert inserts given obj to UniformObjectMap
+func (m UniformObjectMap) Insert(parent v1.Object, obj *unstructured.Unstructured) {
+	internalGvk := GroupVersionKind{obj.GroupVersionKind()}
+	if m[internalGvk] == nil {
+		m.InitGroup(obj.GroupVersionKind())
+	}
+	group := m[internalGvk]
+	name := m.qualifiedName(obj)
+	group[name] = obj
+}
+
+// InsertAll inserts given slice of objects to UniformObjectMap
+func (m UniformObjectMap) InsertAll(parent v1.Object, objects []*unstructured.Unstructured) {
+	for _, object := range objects {
+		m.Insert(parent, object)
+	}
+}
+
+// qualifiedName returns the namespace/name of the object. If obj is clustered scope,
+// return just name.
+func (m UniformObjectMap) qualifiedName(obj *unstructured.Unstructured) string {
+	if obj.GetNamespace() != "" {
+		return fmt.Sprintf("%s/%s", obj.GetNamespace(), obj.GetName())
+	}
+	return obj.GetName()
+}
+
+// FindGroupKindName search object by name in each schema.GroupKind (ignoring version part)
+func (m UniformObjectMap) FindGroupKindName(gk schema.GroupKind, name string) *unstructured.Unstructured {
+	// The map is keyed by group-version-kind, but we don't know the version.
+	// So, check inside any GVK that matches the group and kind, ignoring version.
+	for key, objects := range m {
+		if key.GroupKind() == gk {
+			for n, object := range objects {
```

Comment:
query: wondering if for loops can avoided since m is a map or maps

Reply:
hmm the inner one - I think yes. The outer - the key is GroupVersinKind, I need GroupKind, and I do not now Version